File: src/NodeList/FluidNodeList.cc
//---------------------------------Spheral++----------------------------------//
// FluidNodeList -- An abstract base class for the NodeLists to represent
//                  fluids.
//
// Created by JMO, Sat Sep 18 10:50:42 PDT 1999
//----------------------------------------------------------------------------//
#include "FileIO/FileIO.hh"
#include "Material/EquationOfState.hh"
#include "Hydro/HydroFieldNames.hh"
#include "DataBase/DataBase.hh"
#include "DataBase/IncrementState.hh"
#include "DataBase/ReplaceState.hh"
#include "Kernel/TableKernel.hh"
#include "Field/FieldList.hh"
#include "DataBase/State.hh"
#include "DataBase/StateDerivatives.hh"
#include "Neighbor/ConnectivityMap.hh"
#include "Utilities/safeInv.hh"
#include "FluidNodeList.hh"

using std::vector;
using std::list;
using std::string;
using std::cerr;
using std::endl;

namespace Spheral {

//------------------------------------------------------------------------------
// Construct with given EOS object, along with optional numInternal nodes,
// numGhost nodes, and name.
//------------------------------------------------------------------------------
template<typename Dimension>
FluidNodeList<Dimension>::
FluidNodeList(string name,
              EquationOfState<Dimension>& eos,
              const size_t numInternal,
              const size_t numGhost,
              const Scalar hmin,
              const Scalar hmax,
              const Scalar hminratio,
              const Scalar nPerh,
              const size_t maxNumNeighbors,
              const Scalar rhoMin,
              const Scalar rhoMax):
  NodeList<Dimension>(name, numInternal, numGhost, hmin, hmax, hminratio, nPerh, maxNumNeighbors),
  mRhoMin(rhoMin),
  mRhoMax(rhoMax),
  mMassDensity(HydroFieldNames::massDensity, *this),
  mSpecificThermalEnergy(HydroFieldNames::specificThermalEnergy, *this),
  mEosPtr(&eos) {
}

//------------------------------------------------------------------------------
// Set the mass density field.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
massDensity(const Field<Dimension, typename Dimension::Scalar>& rho) {
  mMassDensity = rho;
  mMassDensity.name(HydroFieldNames::massDensity);
}

//------------------------------------------------------------------------------
// Set the specific thermal energy field.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
specificThermalEnergy(const Field<Dimension, typename Dimension::Scalar>& eps) {
  mSpecificThermalEnergy = eps;
  mSpecificThermalEnergy.name(HydroFieldNames::specificThermalEnergy);
}

//------------------------------------------------------------------------------
// Compute the pressure.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
pressure(Field<Dimension, typename Dimension::Scalar>& field) const {
  REQUIRE(field.nodeListPtr() == this);
  mEosPtr->setPressure(field, mMassDensity, mSpecificThermalEnergy);
  field.name(HydroFieldNames::pressure);
}

//------------------------------------------------------------------------------
// Calculate and return the temperature field.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
temperature(Field<Dimension, typename Dimension::Scalar>& field) const {
  REQUIRE(field.nodeListPtr() == this);
  mEosPtr->setTemperature(field, mMassDensity, mSpecificThermalEnergy);
  field.name(HydroFieldNames::temperature);
}

//------------------------------------------------------------------------------
// Calculate and return the sound speed field.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
soundSpeed(Field<Dimension, typename Dimension::Scalar>& field) const {
  REQUIRE(field.nodeListPtr() == this);
  mEosPtr->setSoundSpeed(field, mMassDensity, mSpecificThermalEnergy);
  field.name(HydroFieldNames::soundSpeed);
}

//------------------------------------------------------------------------------
// Calculate and return the volume per node.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
volume(Field<Dimension, typename Dimension::Scalar>& field) const {
  REQUIRE(field.nodeListPtr() == this);
  const Field<Dimension, Scalar>& nodeMass = this->mass();
  const Field<Dimension, Scalar>& nodeDensity = this->massDensity();
  for (size_t i = 0; i != this->numNodes(); ++i) field[i] = nodeMass[i]*safeInv(nodeDensity[i]);
  field.name(HydroFieldNames::volume);
}

//------------------------------------------------------------------------------
// Calculate and return the linear momentum per node.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
linearMomentum(Field<Dimension, typename Dimension::Vector>& field) const {
  REQUIRE(field.nodeListPtr() == this);
  const Field<Dimension, Scalar>& nodeMass = this->mass();
  const Field<Dimension, Vector>& nodeVelocity = this->velocity();
  for (size_t i = 0; i != this->numNodes(); ++i) field[i] = nodeMass[i]*nodeVelocity[i];
  field.name(HydroFieldNames::linearMomentum);
}

//------------------------------------------------------------------------------
// Calculate and return the total energy per node.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
totalEnergy(Field<Dimension, typename Dimension::Scalar>& field) const {
  REQUIRE(field.nodeListPtr() == this);
  const Field<Dimension, Scalar>& nodeMass = this->mass();
  const Field<Dimension, Vector>& nodeVelocity = this->velocity();
  const Field<Dimension, Scalar>& nodeEps = this->specificThermalEnergy();
  for (size_t i = 0; i != this->numNodes(); ++i) 
    field[i] = nodeMass[i]*(0.5*nodeVelocity[i].magnitude2() + nodeEps[i]);
  field.name(HydroFieldNames::totalEnergy);
}

//------------------------------------------------------------------------------
// Dump the current state of the NodeList to the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
dumpState(FileIO& file, const string& pathName) const {

  // Dump the ancestor class.
  NodeList<Dimension>::dumpState(file, pathName);

  // Dump each of the internal fields of the FluidNodeList.
  file.write(mMassDensity, pathName + "/massDensity");
  file.write(mSpecificThermalEnergy, pathName + "/specificThermalEnergy");
}

//------------------------------------------------------------------------------
// Restore the state of the NodeList from the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
FluidNodeList<Dimension>::
restoreState(const FileIO& file, const string& pathName) {

  // Restore the ancestor class.
  NodeList<Dimension>::restoreState(file, pathName);

  // Restore each of the internal fields of the FluidNodeList.
  file.read(mMassDensity, pathName + "/massDensity");
  file.read(mSpecificThermalEnergy, pathName + "/specificThermalEnergy");
}  

}

File: src/Pybind11Wraps/NodeList/FluidNodeList.py
from PYB11Generator import *
from NodeList import NodeList

#-------------------------------------------------------------------------------
# FluidNodeList template
#-------------------------------------------------------------------------------
@PYB11template("Dimension")
class FluidNodeList(NodeList):
    "Spheral FluidNodeList base class in %(Dimension)s, i.e.,  the NodeList for fluid hydrodynamics."

    def pyinit(self,
               name = "std::string",
               eos = "EquationOfState<%(Dimension)s>&",
               numInternal = ("unsigned", "0"),
               numGhost = ("unsigned", "0"),
               hmin = ("double", "1e-20"),
               hmax = ("double", "1e20"),
               hminratio = ("double", "0.1"),
               nPerh = ("double", "2.01"),
               maxNumNeighbors = ("unsigned", "500"),
               rhoMin = ("double", "1e-10"),
               rhoMax = ("double", "1e100")):
        "Constructor for a FluidNodeList class."
        return

    @PYB11const
    def massDensity(self):
        "The mass density field"
        return "const %(ScalarField)s&"

    @PYB11const
    def specificThermalEnergy(self):
        "The specific thermal energy field"
        return "const %(ScalarField)s&"

    @PYB11virtual
    @PYB11const
    def pressure(self, result="%(ScalarField)s&"):
        "Compute the current pressure, storing the result in the argument %(ScalarField)s."
        return "void"

    @PYB11virtual
    @PYB11const
    def temperature(self, result="%(ScalarField)s&"):
        "Compute the current temperature, storing the result in the argument %(ScalarField)s."
        return "void"

    @PYB11virtual
    @PYB11const
    def soundSpeed(self, result="%(ScalarField)s&"):
        "Compute the current sound speed, storing the result in the argument %(ScalarField)s."
        return "void"

    @PYB11virtual
    @PYB11const
    def volume(self, result="%(ScalarField)s&"):
        "Compute the current volume, storing the result in the argument %(ScalarField)s."
        return "void"

    @PYB11virtual
    @PYB11const
    def linearMomentum(self, result="%(VectorField)s&"):
        "Compute the current linear momentum, storing the result in the argument %(ScalarField)s."
        return "void"

    @PYB11virtual
    @PYB11const
    def totalEnergy(self, result="%(ScalarField)s&"):
        "Compute the current total energy, storing the result in the argument %(ScalarField)s."
        return "void"

    @PYB11const
    def equationOfState(self):
        "Return the equation of state object this FluidNodeList is associated with."
        return "const EquationOfState<%(Dimension)s>&"

    def setequatinoOfState(self, equationOfState="const EquationOfState<%(Dimension)s>&"):
        "Set the equation of state for this FluidNodeList."
        return "void"

    @PYB11virtual
    @PYB11const
    def label(self):
        "Label for restart files"
        return "std::string"

    @PYB11virtual
    @PYB11const
    def dumpState(self, file="FileIO&", pathName="const std::string&"):
        "Serialize under the given path in a FileIO object"
        return "void"

    @PYB11virtual
    def restoreState(self, file="const FileIO&", pathName="const std::string&"):
        "Restore state from the given path in a FileIO object"
        return "void"

    # Comparison
    def __eq__(self):
        "Equivalence test with another FluidNodeList"

    def __ne__(self):
        "Inequivalence test with another FluidNodeList"

    # Methods used for properties
    @PYB11ignore
    @PYB11cppname("rhoMin")
    @PYB11const
    def getrhoMin(self):
        return "double"

    @PYB11ignore
    @PYB11cppname("rhoMin")
    @PYB11const
    def setrhoMin(self, val="double"):
        return "void"

    @PYB11ignore
    @PYB11cppname("rhoMax")
    @PYB11const
    def getrhoMax(self):
        return "double"

    @PYB11ignore
    @PYB11cppname("rhoMax")
    @PYB11const
    def setrhoMax(self, val="double"):
        return "void"

    # Properties
    rhoMin = property(getrhoMin, setrhoMin, doc="The minimum allowed mass density.")
    rhoMax = property(getrhoMax, setrhoMax, doc="The maximum allowed mass density.")
